docs/development_best_practices.md

```diff
@@ -0,0 +1,130 @@
+	* Do one time initialization for live entities in **activate** method, e.g., opening connections to a database server or starting a thread for asynchronous operations. The **activate** method is called right before processing starts so it is a better place to do these kind of initializations than at **setup** which can lead to a delay before processing data from the live entity.  
+	* Perform periodic tasks based on processing time in application windows
+	* Perform initializations needed for each application window in **beginWindow**
+	* Perform aggregations needed for each application window  in **endWindow**
+	* Teardown of live entities (inverse of tasks performed during activate) should be in the **deactivate** method
+	* Teardown of lifetime entities (those initialized in setup method) should happen in the **teardown** method
+	* If the operator implementation is not finalized mark it with the **@Evolving** annotation
+* If the operator needs to perform operations based on event time of the individual tuples and not the processing time, extend and use the **WindowedOperator**. Refer to documentation of that operator for details on how to use it.
+* If an operator needs to do some work when it is not receiving any input, it should implement **IdleTimeHandler** interface. This interface contains **handleIdleTime** method which will be called whenever the platform isn’t doing anything else and the operator can do the work in this method. If for any reason the operator is not going to do any work in this method, it should sleep for a small amount of time such as that specified by the **SPIN_MILLIS** attribute otherwise it will cause a busy wait.
+* Often operators have customizable parameters such as information about locations of external systems or parameters that modify the behavior of the operator. Users should be able to specify these easily without having to change source code. This can be done by making them properties of the operator because they can then be initialized from external properties files.
+	* Where possible default values should be provided for the properties in the source code.
+	* Validation rules should be specified for the properties using javax constraint validations that check whether the values specified for the properties are in the correct format, range or other operator requirements. Required properties should have at least a @NotNull validation specifying that they have to be specified by the user.
+
+### Checkpointing
+
+Checkpointing is a process of snapshotting the state of an operator and saving it so that in case of failure the state can be used to restore the operator and continue processing. It is automatically performed by the platform at a configurable interval. All operators in the application are checkpointed in a distributed fashion, thus allowing the entire state of the application to be saved and available for recovery if needed. Here are some things to remember when it comes to checkpointing
```

Comment:
restore the operator => restore the operator to the prior state
comes to checkpointing => comes to checkpointing: